### ISSUE TITLE
Trouble with Batch Normalization

### ISSUE THREAD

**Issue by slaweks17** [Microsoft/CNTK#790]
Hi,
I tried Batch Normalization and it is not working for me:
While training is seemingly accelerated brilliantly and I get very low training errors quickly, when I do action write and dump the output and calculate error, even on training set, it is very bad, like order of magnitude bigger than that reported during training.
I am doing regression (output is a vector of length 6). The net is a few layer-deep with LRs. The LR layer with BN is defined as follows:

BNReLULayer (inDim, outDim, x, wScale=1, sScale=1) = [  
    W   = Parameter (outDim, inDim, init = "gaussian", initValueScale = wScale, initOnCPUOnly=true) 
    bias   = Parameter (outDim, 1, init = "fixedValue", value = 0) 
    sc  = Parameter (outDim, 1, init = "fixedValue", value = sScale) 
    m   = Parameter (outDim, 1, init = "fixedValue", value = 0, learningRateMultiplier = 0)
    isd = Parameter (outDim, 1, init = "fixedValue", value = 0, learningRateMultiplier = 0)
    t = Times(W, x)  # TODO: W \* x
    bn = BatchNormalization(t, sc, bias, m, isd, false, #754 
      useCntkEngine = true, epsilon = 0.00001)
    y = RectifiedLinear(bn)
].y

and I leave batchNormalizationTimeConstant and batchNormalizationBlendTimeConstant with defaults values.

Doing the same without BN gets me seemingly much higher training errors (SquareError) but much smaller errors when checking the written output.

Regards,
Slawek

**Comment by frankseide** [comment]
Use normalization time constant if 100000

Get Outlook for iOS<https://aka.ms/o0ukef>




On Fri, Aug 26, 2016 at 7:57 AM -0700, "slaweks17" <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi,
I tried Batch Normalization and it is not working for me:
While training is seemingly accelerated brilliantly and I get very low training errors quickly, when I do action write and dump the output and calculate error, even on training set, it is very bad, like order of magnitude bigger than that reported during training.
I am doing regression (output is a vector of length 6). The net is a few layer-deep with LRs. The LR layer with BN is defined as follows:

BNReLULayer (inDim, outDim, x, wScale=1, sScale=1) = [

W = Parameter (outDim, inDim, init = "gaussian", initValueScale = wScale, initOnCPUOnly=true)
bias = Parameter (outDim, 1, init = "fixedValue", value = 0)
sc = Parameter (outDim, 1, init = "fixedValue", value = sScale)
m = Parameter (outDim, 1, init = "fixedValue", value = 0, learningRateMultiplier = 0)
isd = Parameter (outDim, 1, init = "fixedValue", value = 0, learningRateMultiplier = 0)
t = Times(W, x) # TODO: W * x
bn = BatchNormalization(t, sc, bias, m, isd, false, #754<https://github.com/Microsoft/CNTK/issues/754>
useCntkEngine = true, epsilon = 0.00001)
y = RectifiedLinear(bn)
].y

and I leave batchNormalizationTimeConstant and batchNormalizationBlendTimeConstant with defaults values.

Doing the same without BN gets me seemingly much higher training errors (SquareError) but much smaller errors when checking the written output.

Regards,
Slawek

-

**Comment by yuxiaoguo** [comment]
Indeed, your problem do exist in some special circumstance. The mean and isd used in testing phase are from the running average statistics in training phase. Indeed, since the weights updated huge, the statistics is unstable either. Normally, the statistics will achieve better performance while the model is a little stable. If you want to check the test error in training process, I suggest you setting the bnTimeConst to 0. That will use the last mean and std in training as the testing statistics. It may get a better result than yours. However, if you want to get a better result at the end of training, using default bnTimeConst may a good choose.

**Comment by frankseide** [comment]
BatchNorm has a bad default. `normalizationTimeConstant=0` means that the mean/var statistics used in inference is taken from only the very last minibatch of the training. It should be set to something larger. The 1.7 binary, soon to be released, will use a default of 5000, but any value of a few thousand should do.

**Comment by slaweks17** [comment]
Hi,
I experimented quite a bit with various batchNormalizationTimeConstant values, and the best setting seems to be actually 0, followed by something around 1000. Maybe it is because my dataset is small, around 5k rows?
Anyway, I am getting much better results (errors 2,3 times smaller) without Batch Normalization. I will try BN again on larger data sets.

**Comment by slaweks17** [comment]
One more question: does the minibatch size in write action matter in this context?

**Comment by cha-zhang** [comment]
Close this issue since it's very old. Feel free to open a new one if needed.
